### PR TITLE
check for false node  in visit

### PR DESCRIPTION
**Pull request by shmax**
Fix for #685 

A recent [commit](cab1599) introduced a flaw in directive parsing. It seems to have been drawn from the js library, but there's a check for [false](https://github.com/graphql/graphql-js/blob/master/src/language/visitor.js#L308) that apparently didn't get brought over. 

Note: I tried to re-port the entire block in the surrounding code, but the tests blew up. Some day we should re-port the entire visit function, as it has diverged quite a bit.

### CODE REVIEW

**Comment by shmax** [comment]
> Can you check if it works for you?

b110268

It does indeed work, thank you, but I'm a bit puzzled by a few things: 
1. I don't see a KnownArgumentNamesOnDirectives rule in the reference library code
2. All execution paths in `KnownArgumentNamesOnDirectives.getASTVisitor` result in a skip. Is that expected?
3. The reference library _does_ have that check for false. Is it important?

**Comment by shmax** [comment]
> If you can dig up an accompanying test case

There are a few test cases that look relevant, but for some reason it doesn't trigger the bug: https://github.com/webonyx/graphql-php/blob/master/tests/Executor/DirectivesTest.php#L69

I'll see if I can figure it out...

**Comment by vladar** [comment]
> I don't see a KnownArgumentNamesOnDirectives rule in the reference library code

It's inside `KnownArgumentNames` there

**Comment by shmax** [comment]
> It's inside `KnownArgumentNames` there

Ah, so it is. Okay. I don't see any "skip" logic in there, but if you're satisfied, I'm satisfied. Please merge at your convenience.

**Comment by shmax** [comment]
Ack, coverage decreased...

**Comment by vladar** [comment]
> Ah, so it is. Okay. I don't see any "skip" logic in there, but if you're satisfied, I'm satisfied. Please merge at your convenience.

They use `return false` for skip logic here:

https://github.com/graphql/graphql-js/blob/1e1d75e9b7914abb66d91bfa0ae1de8612674628/src/validation/rules/KnownArgumentNamesRule.js#L95

In php library we introduced explicit `Visitor::skipNode()` (and several other static methods) so we deviate a bit from the reference implementation here (and also that's the reason for this error - I just forgot about this, although surprised that tests didn't catch it)

**Comment by shmax** [comment]
> Ack, coverage decreased...

The skip change seems to have resulted in some dead code, which I have removed. 4722ce7

**Comment by shmax** [comment]
> They use return false for skip logic here

Gotcha. That works for me, thanks for the background.

File: src/Language/Visitor.php [hunk 1]
```diff
@@ -417,18 +417,20 @@ public static function visitInParallel($visitors)
...
+                    if (! ($result instanceof VisitorOperation)) {
+                        continue;
+                    }
+
+                    if ($result->doContinue) {
+                        $skipping[$i] = $node;
+                    } elseif ($result->doBreak) {
+                        $skipping[$i] = $result;
+                    } elseif ($result->removeNode) {
                         return $result;
```
Comment (vladar) [inline src/Language/Visitor.php:429]: Looks like we miss this branch now? When the result is not an instance of `VisitorOperation` we must return it

Comment (shmax) [reply src/Language/Visitor.php:429]: Sorry, what is your suggestion?

Comment (shmax) [reply src/Language/Visitor.php:429]: I think I understand what you mean. With your skip change, that bit of code was being flagged as dead by the coverage tests, so I removed it.

Comment (vladar) [reply src/Language/Visitor.php:429]: GitHub shows this comment against new code but I've actually left it against the old code. So I mean this part:

```php
elseif ($result !== null) {
    return $result;
}
```

is missing now. We must restore it as it is basically how you "edit" an AST node with a parallel visitor.

Comment (shmax) [reply src/Language/Visitor.php:432]: Restored in 11b59cd. Are you sure this is necessary? I only ask because the tests passed without it.

Comment (vladar) [reply src/Language/Visitor.php:432]: Absolutely. Probably missing tests for this (we don't use a parallel visitor to modify nodes in this library, yet this feature is indispensable for AST manipulations)

Comment (vladar) [reply src/Language/Visitor.php:432]: Actually, can we revert this change completely (except for `$visitFn !== null` part)? Doesn't look more readable to me after the change.

Comment (shmax) [reply src/Language/Visitor.php:432]: Wait, what? Revert what change? The one I just made?

Comment (vladar) [reply src/Language/Visitor.php:432]: I mean all changes in this file except for `if ($visitFn !== null) {` line. Those changes don't really change anything now that we restored `return $result`. Or am I missing something?

Comment (shmax) [reply src/Language/Visitor.php]: Sure, you got it 7e680fd

**Comment by shmax** [comment]
It seems coverage has decreased a bit:

https://coveralls.io/builds/31804139/source?filename=src%2FLanguage%2FVisitor.php#L429

@vladar  Can I ask your assistance in doing whatever it takes to get this PR merged? Days are rolling by, and this problem is blocking other developers.

**Comment by shmax** [comment]
Thanks @vladar and @spawnia for reviewing!

**Comment by vladar** [comment]
Published in `v14.0.1`

**Pull request merged**